### ISSUE TITLE
Expected dereference.of.nullable error not found by using nullness checker in nested class method

### ISSUE THREAD

**Issue by AndrewShf** [eisop/checker-framework#226]
A false negative found in using the nullness checker:
Input:
```
public class InitializationTest {

    {
        Object o =
                new Object() {
                    @Override
                    public String toString() {
                        // :: error: (dereference.of.nullable)
                        return a.toString();
                    }
                }.toString();
    }
    String a = "";

    public static void main(String[] args) {new InitializationTest();}
}
```

Command:
`javac InitializationTest.java && java InitializationTest`

Output:
```
Exception in thread "main" java.lang.NullPointerException
	at InitializationTest$1.toString(InitializationTest.java:9)
	at InitializationTest.<init>(InitializationTest.java:11)
	at InitializationTest.main(InitializationTest.java:15)
```

Run Nullness cheker:
`javacheck -processor org.checkerframework.checker.nullness.NullnessChecker InitializationTest.java`

Output:
No error reported (Expects `dereference.of.nullable` on line 9).

**Comment by AndrewShf** [comment]
One interesting thing to notice: if adding `final` to the declaration of field `a`, no error will occur in running this program
```
public class InitializationTest {

    {
        Object o =
                new Object() {
                    @Override
                    public String toString() {
                        // :: error: (dereference.of.nullable)
                        return a.toString();
                    }
                }.toString();
    }
    final String a = "";

    public static void main(String[] args) {new InitializationTest();}
}
```
Again run `javac InitializationTest.java && java InitializationTest`. No error reported.

So maybe adding `final` modifier changes the sequence of initialization?
